### ISSUE TITLE
Add non shader visible descriptor pool

### ISSUE THREAD

**Issue by 01Pollux** [NVIDIAGameWorks/NRI#53]
It would be very useful to have a cpy only descriptor for copying and storages.

**Comment by dzhdanNV** [comment]
Please, elaborate on your request. What functionality is needed? Can you summarize your request into a draft of API changes?

**Comment by 01Pollux** [comment]
Well what i meant by that was, the developer can allocate a descriptor heap without `D3D12_DESCRIPTOR_HEAP_FLAG_SHADER_VISIBLE` flag in D3D12 or with `VK_DESCRIPTOR_POOL_CREATE_HOST_ONLY_BIT_EXT` in vulkan. then create resource views from them (for example in advance during app initialization).
When we would need the resource in the shader, it would be faster to copy it from non-shader visbile descriptor to shader visbile descriptor rather than recreate the same view, or copy it from shader visible descriptor.
The changes im proposing is to add either `isHostDescriptor` variable to `DescriptorPoolDesc` during pool creation and delegate the views management to the developer.
I apologize if im not clear enough.

**Comment by dzhdanNV** [comment]
Thanks for the details. Additional question - what's the problem with the current solution? Sorry, I'm just trying to understand all the details.

**Comment by 01Pollux** [comment]
The current implementation doesn't allow rooms for optimizations, such as caching ahead the descriptor and only uploading it online when needed. It could be done with current implementation, but offline descriptors are made for these kind of operations.

**Comment by 01Pollux** [comment]
> Current NRI:
> 
> * `CreateX(..., Descriptor** )` - creates a CPU-only descriptor
> * when you `Update` a `DescriptorSet` these CPU-only views get copied into a GPU-visible `DescriptorPool`
> * `DescriptorPool` are always GPU-visible

That is it, I suppose it already does what my previous comments were aiming for. I apologize for the confusion.


> But nevertheless, changes are minimal and your proposal seems helpful, so it is worth discussing. You suggest adding `bool hostOnly` to `DescriptorPoolDesc` and proper handling of `SHADER_VISIBLE` / `HOST_ONLY` in D3D12/VK, right? AFAIU, that's it - no more changes needed. A user will be able to use already existing NRI API to create `DescriptorSets` from classic (`SHADER_VISIBLE`) and proposed (`HOST_ONLY`) pools and copy between them. All combinations should work:
> 
> * `SHADER_VISIBLE` descriptor set => SHADER_VISIBLE descriptor set (classic)
> * `HOST_ONLY` descriptor set => `SHADER_VISIBLE` descriptor set (proposal)

Sounds good to have custom allocator for descriptor pool.

**Comment by dzhdanNV** [comment]
I will add `bool hostOnly` to `DescriptorPoolDesc`. We will start from here. It will be included into the next release (next week I hope).

**Comment by dzhdanNV** [comment]
The benefit is unclear: in D3D12 we already manage non-shader visible heaps under the hood. This change will allow you to "introduce" a redundant copy via `UpdateDescriptorRanges` from one non-shader visible (internal) heap to another non-shader visible (user) heap to be able to use sets from this heap as "copy source" later via `CopyDescriptorSet`, with the only one benefit - layout change (i.e. order of descriptors). Is it really needed?

**Comment by 01Pollux** [comment]
> The benefit is unclear: in D3D12 we already manage non-shader visible heaps under the hood. This change will allow you to "introduce" a redundant copy via `UpdateDescriptorRanges` from one non-shader visible (internal) heap to another non-shader visible (user) heap to be able to use sets from this heap as "copy source" later via `CopyDescriptorSet`, with the only one benefit - layout change (i.e. order of descriptors). Is it really needed?

After careful consideration, I have decided not to pursue the idea any further. Thanks for clarifying my misconception.

**Comment by dzhdanNV** [comment]
Closing? (at least for now)

**Comment by dzhdanNV** [comment]
Thanks. Feel free to reopen in the future.
